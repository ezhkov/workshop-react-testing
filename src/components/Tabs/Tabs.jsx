import React from 'react';
import {
  Tab, Tabs as TabsContainer, TabList, TabPanel,
} from 'react-tabs';
import nanoid from 'nanoid';

import 'react-tabs/style/react-tabs.css';

export default class Tabs extends React.Component {
  state = {
    tabs: [
      {
        title: 'Tab header 1',
        content: 'Tab panel 1 content',
        uid: nanoid(),
      },
      {
        title: 'Tab header 2',
        content: 'Tab panel 2 content',
        uid: nanoid(),
      },
      {
        title: 'Tab header 3',
        content: 'Tab panel 3 content',
        uid: nanoid(),
      },
      {
        title: 'Tab header 4',
        content: 'Tab panel 4 content',
        uid: nanoid(),
      },
    ],
  };

  makeTab = () => {
    const { tabs } = this.state;
    return {
      title: `Tab header ${tabs.length + 1}`,
      content: `Tab panel ${tabs.length + 1} content`,
      uid: nanoid(),
    };
  };

  handleRemoveTab = index => () => {
    const { activeTabIndex } = this.state;
    if (index !== activeTabIndex) {
      this.setState((prevState) => {
        const newTabs = [...prevState.tabs];
        newTabs.splice(index, 1);
        return {
          tabs: newTabs,
        };
      });
    }
  };

  handleAddTab = () => {
    const newTab = this.makeTab();
    this.setState(prevState => ({
      tabs: [
        ...prevState.tabs,
        newTab,
      ],
    }));
  };

  renderTabsList() {
    const { tabs } = this.state;
    return tabs.map((elem, index) => (
      <Tab key={elem.uid} data-test="tab">
        <button type="button" data-test="removeTab" onClick={this.handleRemoveTab(index)}>X</button>
        <span>{ elem.title }</span>
      </Tab>
    ));
  }

  renderTabsPanels() {
    const { tabs } = this.state;
    return tabs.map(elem => (
      <TabPanel key={elem.uid} data-test="tabContent">
        <div>
          { elem.content }
        </div>
      </TabPanel>
    ));
  }

  render() {
    return (
      <>
        <div>
          <button data-test="addTab" type="button" onClick={this.handleAddTab}>Add tab</button>
        </div>
        <TabsContainer>
          <TabList data-test="tabs-box">
            { this.renderTabsList() }
          </TabList>
          { this.renderTabsPanels() }
        </TabsContainer>
      </>
    );
  }
}
