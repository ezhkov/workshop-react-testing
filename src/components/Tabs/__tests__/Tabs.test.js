import React from 'react';
import Tabs from '..';
import { shallow, mount } from 'enzyme';

describe('Tabs', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Tabs />);
    expect(wrapper).toMatchSnapshot();
  });

  it('select 3rd tab', () => {
    const wrapper = mount(<Tabs />);
    const elemToClick = wrapper.find('Tab').at(2);
    elemToClick.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});
