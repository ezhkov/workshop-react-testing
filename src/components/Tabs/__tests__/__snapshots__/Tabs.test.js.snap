// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders correctly 1`] = `
<Tabs
  defaultFocus={false}
  defaultIndex={null}
  forceRenderTabPanel={false}
  selectedIndex={null}
>
  <TabList
    className="react-tabs__tab-list"
  >
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Mario
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabled={true}
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Luigi
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Peach
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Yoshi
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Toad
    </Tab>
  </TabList>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <p>
      <b>
        Mario
      </b>
       
      (
      <i>
        Japanese: マリオ Hepburn: Mario, [ma.ɾʲi.o]
      </i>
      ) (
      <i>
        English: /ˈmɑːrioʊ/; Italian: [ˈmaːrjo]
      </i>
      ) is a fictional character in the Mario video game franchise, owned by Nintendo and created by Japanese video game designer Shigeru Miyamoto. Serving as the company's mascot and the eponymous protagonist of the series, Mario has appeared in over 200 video games since his creation. Depicted as a short, pudgy, Italian plumber who resides in the Mushroom Kingdom, his adventures generally center upon rescuing Princess Peach from the Koopa villain Bowser. His younger brother and sidekick is Luigi.
    </p>
    <p>
      Source:
       
      <a
        href="https://en.wikipedia.org/wiki/Mario"
        target="_blank"
      >
        Wikipedia
      </a>
    </p>
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <p>
      <b>
        Luigi
      </b>
       
      (
      <i>
        Japanese: ルイージ Hepburn: Ruīji, [ɾɯ.iː.dʑi̥]
      </i>
      ) (
      <i>
        English: /luˈiːdʒi/; Italian: [luˈiːdʒi]
      </i>
      ) is a fictional character featured in video games and related media released by Nintendo. Created by prominent game designer Shigeru Miyamoto, Luigi is portrayed as the slightly younger but taller fraternal twin brother of Nintendo's mascot Mario, and appears in many games throughout the Mario franchise, often as a sidekick to his brother.
    </p>
    <p>
      Source:
       
      <a
        href="https://en.wikipedia.org/wiki/Luigi"
        target="_blank"
      >
        Wikipedia
      </a>
    </p>
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <p>
      <b>
        Princess Peach
      </b>
       
      (
      <i>
        Japanese: ピーチ姫 Hepburn: Pīchi-hime, [piː.tɕi̥ çi̥.me]
      </i>
      ) is a character in Nintendo's Mario franchise. Originally created by Shigeru Miyamoto, Peach is the princess of the fictional Mushroom Kingdom, which is constantly under attack by Bowser. She often plays the damsel in distress role within the series and is the lead female. She is often portrayed as Mario's love interest and has appeared in Super Princess Peach, where she is the main playable character.
    </p>
    <p>
      Source:
       
      <a
        href="https://en.wikipedia.org/wiki/Princess_Peach"
        target="_blank"
      >
        Wikipedia
      </a>
    </p>
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <p>
      <b>
        Yoshi
      </b>
       
      (
      <i>
        ヨッシー Yosshī, [joɕ.ɕiː]
      </i>
      ) (
      <i>
        English: /ˈjoʊʃi/ or /ˈjɒʃi/
      </i>
      ), once romanized as Yossy, is a fictional anthropomorphic dinosaur who appears in video games published by Nintendo. Yoshi debuted in Super Mario World (1990) on the Super Nintendo Entertainment System as Mario and Luigi's sidekick. Yoshi later starred in platform and puzzle games, including Super Mario World 2: Yoshi's Island, Yoshi's Story and Yoshi's Woolly World. Yoshi also appears in many of the Mario spin-off games, including Mario Party and Mario Kart, various Mario sports games, and Nintendo's crossover fighting game series Super Smash Bros. Yoshi belongs to the species of the same name, which is characterized by their variety of colors.
    </p>
    <p>
      Source:
       
      <a
        href="https://en.wikipedia.org/wiki/Yoshi"
        target="_blank"
      >
        Wikipedia
      </a>
    </p>
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <p>
      <b>
        Toad
      </b>
       
      (
      <i>
        Japanese: キノピオ Hepburn: Kinopio
      </i>
      ) is a fictional character who primarily appears in Nintendo's Mario franchise. Created by Japanese video game designer Shigeru Miyamoto, he is portrayed as a citizen of the Mushroom Kingdom and is one of Princess Peach's most loyal attendants; constantly working on her behalf. He is usually seen as a non-player character (NPC) who provides assistance to Mario and his friends in most games, but there are times when Toad(s) takes center stage and appears as a protagonist, as seen in Super Mario Bros. 2, Wario's Woods, Super Mario 3D World, and Captain Toad: Treasure Tracker.
    </p>
    <p>
      Source:
       
      <a
        href="https://en.wikipedia.org/wiki/Toad_(Nintendo)"
        target="_blank"
      >
        Wikipedia
      </a>
    </p>
  </TabPanel>
</Tabs>
`;

exports[`Tabs select 3rd tab 1`] = `
<Tabs>
  <Tabs
    defaultFocus={false}
    defaultIndex={null}
    forceRenderTabPanel={false}
    selectedIndex={null}
  >
    <UncontrolledTabs
      className="react-tabs"
      focus={false}
      forceRenderTabPanel={false}
      onSelect={[Function]}
      selectedIndex={2}
    >
      <div
        className="react-tabs"
        data-tabs={true}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <TabList
          className="react-tabs__tab-list"
          key=".0"
        >
          <ul
            className="react-tabs__tab-list"
            role="tablist"
          >
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-0"
              key=".0"
              panelId="react-tabs-1"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-1"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-0"
                role="tab"
                tabIndex={null}
              >
                Mario
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabled={true}
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-2"
              key=".1"
              panelId="react-tabs-3"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-3"
                aria-disabled="true"
                aria-selected="false"
                className="react-tabs__tab react-tabs__tab--disabled"
                id="react-tabs-2"
                role="tab"
                tabIndex={null}
              >
                Luigi
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-4"
              key=".2"
              panelId="react-tabs-5"
              selected={true}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-5"
                aria-disabled="false"
                aria-selected="true"
                className="react-tabs__tab react-tabs__tab--selected"
                id="react-tabs-4"
                role="tab"
                tabIndex="0"
              >
                Peach
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-6"
              key=".3"
              panelId="react-tabs-7"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-7"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-6"
                role="tab"
                tabIndex={null}
              >
                Yoshi
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-8"
              key=".4"
              panelId="react-tabs-9"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-9"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-8"
                role="tab"
                tabIndex={null}
              >
                Toad
              </li>
            </Tab>
          </ul>
        </TabList>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-1"
          key=".1"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-0"
        >
          <div
            aria-labelledby="react-tabs-0"
            className="react-tabs__tab-panel"
            id="react-tabs-1"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-3"
          key=".2"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-2"
        >
          <div
            aria-labelledby="react-tabs-2"
            className="react-tabs__tab-panel"
            id="react-tabs-3"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-5"
          key=".3"
          selected={true}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-4"
        >
          <div
            aria-labelledby="react-tabs-4"
            className="react-tabs__tab-panel react-tabs__tab-panel--selected"
            id="react-tabs-5"
            role="tabpanel"
          >
            <p>
              <b>
                Princess Peach
              </b>
               
              (
              <i>
                Japanese: ピーチ姫 Hepburn: Pīchi-hime, [piː.tɕi̥ çi̥.me]
              </i>
              ) is a character in Nintendo's Mario franchise. Originally created by Shigeru Miyamoto, Peach is the princess of the fictional Mushroom Kingdom, which is constantly under attack by Bowser. She often plays the damsel in distress role within the series and is the lead female. She is often portrayed as Mario's love interest and has appeared in Super Princess Peach, where she is the main playable character.
            </p>
            <p>
              Source:
               
              <a
                href="https://en.wikipedia.org/wiki/Princess_Peach"
                target="_blank"
              >
                Wikipedia
              </a>
            </p>
          </div>
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-7"
          key=".4"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-6"
        >
          <div
            aria-labelledby="react-tabs-6"
            className="react-tabs__tab-panel"
            id="react-tabs-7"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-9"
          key=".5"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-8"
        >
          <div
            aria-labelledby="react-tabs-8"
            className="react-tabs__tab-panel"
            id="react-tabs-9"
            role="tabpanel"
          />
        </TabPanel>
      </div>
    </UncontrolledTabs>
  </Tabs>
</Tabs>
`;

exports[`Tabs select disabled tab 1`] = `
<Tabs>
  <Tabs
    defaultFocus={false}
    defaultIndex={null}
    forceRenderTabPanel={false}
    selectedIndex={null}
  >
    <UncontrolledTabs
      className="react-tabs"
      focus={false}
      forceRenderTabPanel={false}
      onSelect={[Function]}
      selectedIndex={0}
    >
      <div
        className="react-tabs"
        data-tabs={true}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <TabList
          className="react-tabs__tab-list"
          key=".0"
        >
          <ul
            className="react-tabs__tab-list"
            role="tablist"
          >
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-0"
              key=".0"
              panelId="react-tabs-1"
              selected={true}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-1"
                aria-disabled="false"
                aria-selected="true"
                className="react-tabs__tab react-tabs__tab--selected"
                id="react-tabs-0"
                role="tab"
                tabIndex="0"
              >
                Mario
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabled={true}
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-2"
              key=".1"
              panelId="react-tabs-3"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-3"
                aria-disabled="true"
                aria-selected="false"
                className="react-tabs__tab react-tabs__tab--disabled"
                id="react-tabs-2"
                role="tab"
                tabIndex={null}
              >
                Luigi
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-4"
              key=".2"
              panelId="react-tabs-5"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-5"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-4"
                role="tab"
                tabIndex={null}
              >
                Peach
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-6"
              key=".3"
              panelId="react-tabs-7"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-7"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-6"
                role="tab"
                tabIndex={null}
              >
                Yoshi
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-8"
              key=".4"
              panelId="react-tabs-9"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-9"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-8"
                role="tab"
                tabIndex={null}
              >
                Toad
              </li>
            </Tab>
          </ul>
        </TabList>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-1"
          key=".1"
          selected={true}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-0"
        >
          <div
            aria-labelledby="react-tabs-0"
            className="react-tabs__tab-panel react-tabs__tab-panel--selected"
            id="react-tabs-1"
            role="tabpanel"
          >
            <p>
              <b>
                Mario
              </b>
               
              (
              <i>
                Japanese: マリオ Hepburn: Mario, [ma.ɾʲi.o]
              </i>
              ) (
              <i>
                English: /ˈmɑːrioʊ/; Italian: [ˈmaːrjo]
              </i>
              ) is a fictional character in the Mario video game franchise, owned by Nintendo and created by Japanese video game designer Shigeru Miyamoto. Serving as the company's mascot and the eponymous protagonist of the series, Mario has appeared in over 200 video games since his creation. Depicted as a short, pudgy, Italian plumber who resides in the Mushroom Kingdom, his adventures generally center upon rescuing Princess Peach from the Koopa villain Bowser. His younger brother and sidekick is Luigi.
            </p>
            <p>
              Source:
               
              <a
                href="https://en.wikipedia.org/wiki/Mario"
                target="_blank"
              >
                Wikipedia
              </a>
            </p>
          </div>
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-3"
          key=".2"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-2"
        >
          <div
            aria-labelledby="react-tabs-2"
            className="react-tabs__tab-panel"
            id="react-tabs-3"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-5"
          key=".3"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-4"
        >
          <div
            aria-labelledby="react-tabs-4"
            className="react-tabs__tab-panel"
            id="react-tabs-5"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-7"
          key=".4"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-6"
        >
          <div
            aria-labelledby="react-tabs-6"
            className="react-tabs__tab-panel"
            id="react-tabs-7"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-9"
          key=".5"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-8"
        >
          <div
            aria-labelledby="react-tabs-8"
            className="react-tabs__tab-panel"
            id="react-tabs-9"
            role="tabpanel"
          />
        </TabPanel>
      </div>
    </UncontrolledTabs>
  </Tabs>
</Tabs>
`;
